### ISSUE TITLE
mrconvert: Segfault

### ISSUE THREAD

**Issue by Lestropie** [jdtournier/mrtrix3#196]
Running with `-zero` and `axes 0,1,2` together in the one call on a 4D image segfaults.

On a related note: Never been a fan of the code branching in mrconvert. Could the `-zero` option be handled in `mrcalc` instead?

**Comment by jdtournier** [comment]
Can't reproduce. Are you sure this is a clean build...?

And I know what you mean re the code branching, it's not pretty. I guess I wouldn't mind shunting it over to `mrcalc`, it admittedly is a better place for it...

**Comment by Lestropie** [comment]
Not really sure. Was on our HPCs, have had weird installation issues there before. Either option alone was happy, but run them together and unhappiness ensued. Really more as a self reminder; should have self-assigned...

Have wanted that functionality out of `mrconvert` for a while, just never been 100% sure how to do it. Could have a `-nan2zero` operator in `mrcalc`, avoid having to do a `-finite` then `-mult`? (I;m guessing the explicit option is there currently because it's used a lot?) Or more generally, a `-replace` ternary operator that does C->A if A==B?

**Comment by jdtournier** [comment]
OK, I won't worry about it then.

I don't think the option is actually used much, if at all... There was some reason I included it in the distant past (it's in the 0.2.x series), but to be honest I doubt anyone will miss it. Happy to just remove it. 

And the syntax for `mrcalc` is fine as-is. As long as people can figure out how to do it, I don't think we need to do anything about it. But it does remind me that we probably need to add `-dup` and `-exch` options in `mrcalc` to duplicate and exchange stack entries respectively. This would avoid the need to load the image twice in the call, i.e. the syntax would change from:

```ShellSession
$ mrcalc in.mif in.mif -finite -mult out.mif
```

to: 

```ShellSession
$ mrcalc in.mif -dup -finite -mult out.mif
```

Would also be useful to if statement, etc. But then, I have a strong suspicion people will probably just supply the argument twice, without looking at this possibility. So maybe that would just be a waste of time. Might be better to try to detect when the same image is being loaded twice...?

**Comment by Lestropie** [comment]
Might have been added because back then there were separate commands for e.g. `mradd`, `mrmult`, no `mrcalc` ... If nothing is jogging your memory, I'd also be tempted to remove `-zero` from `mrconvert`. Potential advantage of reducing the code branching in `mrconvert` is doing the conversion in the output image datatype, rather than converting everything that goes through `mrconvert` to complex float.

Personally I would never use `-dup`, stack-based syntax is tough enough as it is. And I suspect most people would do:
`mrcalc in.mif -finite in.mif -mult out.mif`
rather than
`mrcalc in.mif in.mif -finite -mult out.mif`
anyways; myself included. And trying to avoid double image load is probably more trouble than it's worth. But if you think you'd use it, and if you have a use case for `exch`, there's no harm in them being there.

**Comment by jdtournier** [comment]
OK, I'll have a crack at removing the `-zero` option from `mrconvert`...

**Comment by jdtournier** [comment]
closed with commit 37abeee. Also had a go at cleaning up some other stuff, and implementing the changes to `mrcalc` mentioned above (prevent multiple loads of the same image - see commit 83fe09a).

**Comment by Lestropie** [comment]
Thought after writing that preventing multiple image loads in `mrcalc` might not be too hard after all. Looks like that was the case.

Also: e10153df. Couldn't help myself.
